src/producer/base_producer.rs

```diff
@@ -494,26 +499,22 @@ where
     // As this library uses the rdkafka Event API, flush will not call rd_kafka_poll() but instead wait for
     // the librdkafka-handled message count to reach zero. Runs until value reaches zero or timeout.
     fn flush<T: Into<Timeout>>(&self, timeout: T) -> KafkaResult<()> {
-        let mut timeout = timeout.into();
-        loop {
-            let op_timeout = std::cmp::min(timeout, self.min_poll_interval);
-            if self.in_flight_count() > 0 {
-                unsafe { rdsys::rd_kafka_flush(self.native_ptr(), 0) };
-                self.poll(op_timeout);
+        let deadline: Deadline = timeout.into().into();
+        while self.in_flight_count() > 0 && !deadline.elapsed() {
+            let ret = unsafe {
+                // This cast to i32 will truncate to i32::MAX
+                rdsys::rd_kafka_flush(
+                    self.client().native_ptr(),
+                    deadline.remaining().as_millis() as i32,
```

Comment:
maybe `Deadline` could have a dedicated method instead like `remaining_millis_i32` and inside it we could deal with everything related to conversion? 

We could avoid having https://github.com/fede1024/rust-rdkafka/pull/748/files#diff-4007187965c1fea4fd252a76f5f1007b2ea59f6d86e2f454edee547c5036be24R113 and inside `remaining_millis_i32` do the necessary clamping.

Currently it's clunky as we assume here that `remaining()` or `Deadline` in general is doing the clamping under the hood.

Reply:
Fixed in 68ed9061.